{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0134f159",
   "metadata": {
    "tags": [
     "npsphinx-thumbnail"
    ]
   },
   "source": [
    "# Map single cells to Space (Slide-seqV2 Mouse Hippocampus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f628da43",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from time import time\n",
    "\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "from SC2Spa import SI, PP, Vis, SVA\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from numpy.random import seed\n",
    "from tensorflow.random import set_seed\n",
    "import tensorflow as tf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74ade01d",
   "metadata": {},
   "source": [
    "## Download datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4d1f269e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-01-08 18:35:56--  https://figshare.com/ndownloader/files/38736651\n",
      "Resolving figshare.com (figshare.com)... 63.35.35.68, 63.32.177.41, 2a05:d018:1f4:d003:a6c:2d91:83f8:9cfb, ...\n",
      "Connecting to figshare.com (figshare.com)|63.35.35.68|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/38736651/AdataMH1.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230108/eu-west-1/s3/aws4_request&X-Amz-Date=20230108T173557Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=cb594a3e00979ef0d36ad2db45bf5caeaa843eac9351c468b3a769f3dda98b85 [following]\n",
      "--2023-01-08 18:35:57--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/38736651/AdataMH1.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230108/eu-west-1/s3/aws4_request&X-Amz-Date=20230108T173557Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=cb594a3e00979ef0d36ad2db45bf5caeaa843eac9351c468b3a769f3dda98b85\n",
      "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.1.99, 52.218.110.43, 52.92.33.32, ...\n",
      "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.1.99|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 73481745 (70M) [application/octet-stream]\n",
      "Saving to: ‘Dataset/AdataMH1.h5ad’\n",
      "\n",
      "Dataset/AdataMH1.h5 100%[===================>]  70.08M  11.3MB/s    in 6.3s    \n",
      "\n",
      "2023-01-08 18:36:03 (11.1 MB/s) - ‘Dataset/AdataMH1.h5ad’ saved [73481745/73481745]\n",
      "\n",
      "--2023-01-08 18:36:03--  https://figshare.com/ndownloader/files/38738136\n",
      "Resolving figshare.com (figshare.com)... 63.32.177.41, 63.35.35.68, 2a05:d018:1f4:d000:bec2:d3b8:812e:6b90, ...\n",
      "Connecting to figshare.com (figshare.com)|63.32.177.41|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/38738136/AMB_HC.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230108/eu-west-1/s3/aws4_request&X-Amz-Date=20230108T173603Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=3d4d36ca6981851ea31cd65f002b583e4779105798742f4e77639e5578b5e396 [following]\n",
      "--2023-01-08 18:36:03--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/38738136/AMB_HC.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230108/eu-west-1/s3/aws4_request&X-Amz-Date=20230108T173603Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=3d4d36ca6981851ea31cd65f002b583e4779105798742f4e77639e5578b5e396\n",
      "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.36.10, 52.218.24.26, 52.218.65.28, ...\n",
      "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.36.10|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 334743216 (319M) [application/octet-stream]\n",
      "Saving to: ‘Dataset/AMB_HC.h5ad’\n",
      "\n",
      "Dataset/AMB_HC.h5ad 100%[===================>] 319.24M  11.3MB/s    in 28s     \n",
      "\n",
      "2023-01-08 18:36:32 (11.2 MB/s) - ‘Dataset/AMB_HC.h5ad’ saved [334743216/334743216]\n",
      "\n",
      "--2023-01-08 18:36:32--  https://figshare.com/ndownloader/files/38756529\n",
      "Resolving figshare.com (figshare.com)... 63.32.177.41, 63.35.35.68, 2a05:d018:1f4:d000:bec2:d3b8:812e:6b90, ...\n",
      "Connecting to figshare.com (figshare.com)|63.32.177.41|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/38756529/ssHippo_RCTD.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230108/eu-west-1/s3/aws4_request&X-Amz-Date=20230108T173632Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=a54cd2b86da6e1acc7ac09cef8edd9b68e368fb8c5b11577437d6e07adfd852b [following]\n",
      "--2023-01-08 18:36:32--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/38756529/ssHippo_RCTD.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230108/eu-west-1/s3/aws4_request&X-Amz-Date=20230108T173632Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=a54cd2b86da6e1acc7ac09cef8edd9b68e368fb8c5b11577437d6e07adfd852b\n",
      "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.117.176, 52.218.98.123, 52.218.93.11, ...\n",
      "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.117.176|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1357460 (1.3M) [text/csv]\n",
      "Saving to: ‘Dataset/ssHippo_RCTD.csv’\n",
      "\n",
      "Dataset/ssHippo_RCT 100%[===================>]   1.29M  6.22MB/s    in 0.2s    \n",
      "\n",
      "2023-01-08 18:36:33 (6.22 MB/s) - ‘Dataset/ssHippo_RCTD.csv’ saved [1357460/1357460]\n",
      "\n",
      "/mnt/win1/SC2Spa_Notebook/tutorial1\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists('Dataset'):\n",
    "    os.makedirs('Dataset')\n",
    "!wget https://figshare.com/ndownloader/files/38736651 -O Dataset/AdataMH1.h5ad\n",
    "!wget https://figshare.com/ndownloader/files/38738136 -O Dataset/AMB_HC.h5ad\n",
    "!wget https://figshare.com/ndownloader/files/38756529 -O Dataset/ssHippo_RCTD.csv\n",
    "\n",
    "if not os.path.exists('tutorial1'):\n",
    "    os.makedirs('tutorial1')\n",
    "%cd tutorial1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eebdac00",
   "metadata": {},
   "source": [
    "## Load datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c37af7e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/tmp/ipykernel_56444/677049553.py:30: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  adata_query.obs['common_name'] = adata_query.obs['common_name'].str.replace('?', '')\n"
     ]
    }
   ],
   "source": [
    "#Load\n",
    "adata_ref = ad.read_h5ad('../Dataset/AdataMH1.h5ad')\n",
    "adata_query = ad.read_h5ad('../Dataset/AMB_HC.h5ad')\n",
    "\n",
    "adata_ref.var_names = adata_ref.var_names.str.upper()\n",
    "adata_query.var_names = adata_query.var_names.str.upper()\n",
    "\n",
    "adata_ref.var_names_make_unique()\n",
    "adata_query.var_names_make_unique()\n",
    "\n",
    "#Normalize\n",
    "sc.pp.normalize_total(adata_ref, target_sum=1e4)\n",
    "sc.pp.log1p(adata_ref)\n",
    "sc.pp.normalize_total(adata_query, target_sum=1e4)\n",
    "sc.pp.log1p(adata_query)\n",
    "\n",
    "#Load annotation\n",
    "Anno = pd.read_csv('../Dataset/ssHippo_RCTD.csv', index_col = 0)\n",
    "Anno['MCT'] = 't'\n",
    "index1 = Anno.index[(Anno['celltype_1'] == Anno['celltype_2'])]\n",
    "Anno['MCT'][index1] = Anno['celltype_1'][index1]\n",
    "index2 = Anno.index[(Anno['celltype_1'] != Anno['celltype_2'])]\n",
    "Anno['MCT'][index2] = (Anno['celltype_1'][index2] + '_' + Anno['celltype_2'][index2]).apply(lambda x: '_'.join(sorted(set(x.split('_')))))\n",
    "adata_ref.obs = adata_ref.obs.merge(Anno, left_index = True, right_index = True, how = 'left')\n",
    "\n",
    "adata_ref.obsm['spatial'] = adata_ref.obs[['xcoord', 'ycoord']].values\n",
    "\n",
    "\n",
    "\n",
    "adata_query.obs['common_name'] = adata_query.obs['common_name'].str.replace('?', '')\n",
    "adata_query.obs['simp_name'] = adata_query.obs['common_name'].str.split('.',\n",
    "                        expand = True)[0].str.split(',', expand = True)[0].str.split(' \\(',\n",
    "                                    expand = True)[0].str.replace('cortexm', 'cortex').replace('Medial entorrhinal cortex', 'Medial entorhinal cortex')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fa74a3b",
   "metadata": {},
   "source": [
    "## Select genes using Wasserstein distance (Optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "600db87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sta = time()\n",
    "JGs, WDs = SI.WassersteinD(adata_ref, adata_query, sparse = True,\n",
    "                           WD_cutoff = 0.1, root = 'WDs/', save = 'WDs_T2')\n",
    "\n",
    "end = time()\n",
    "print((end - sta) / 60.0, 'min')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03a7abad",
   "metadata": {},
   "outputs": [],
   "source": [
    "WD_cutoff = 0.4\n",
    "\n",
    "root = 'WDs/'\n",
    "save = 'WDs_T2'\n",
    "\n",
    "WDs = pd.read_csv(root + save + '.csv')\n",
    "JGs = sorted(WDs[WDs['Wasserstein_Distance'] < WD_cutoff]['Gene'].tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de4ccb17",
   "metadata": {},
   "source": [
    "## Fine Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f4aae48",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set random generator seed\n",
    "seed_num = 2022\n",
    "seed(seed_num)\n",
    "set_seed(seed_num)\n",
    "tf.keras.utils.set_random_seed(seed_num)\n",
    "\n",
    "'''\n",
    "Finely map single cells to spatial locations.\n",
    "A model will be trained and saved to `root+name+'.h5'` if model_path is None and save is True.\n",
    "The predicted coordinates of single cells will be saved in adata_query.obsm['spatial_mapping']\n",
    "The predicted coordinates of beads will be saved in adata_ref.obsm['spatial_mapping']\n",
    "Fine mapping information will be saved in adata_ref.obs['FM'] and adata_query.obs['FM']. True if a cell/bead\n",
    "was mapped, otherwise False.\n",
    "'''\n",
    "sta = time()\n",
    "\n",
    "neighbors, dis = SI.FineMapping(adata_ref, adata_query, sparse =True, JGs = None, \n",
    "                                model_path = None, root = 'Model_SI/',\n",
    "                                name = 'SI_T2', l1_reg = 1e-5, l2_reg = 0, dropout = 0.05, epoch = 500,\n",
    "                                batch_size = 4096, nodes = [4096, 1024, 256, 64, 16, 4], lrr_patience = 20,\n",
    "                                ES_patience = 50, min_lr = 1e-5, save = True, polar = True,\n",
    "                                n_neighbors = 1000, dis_cutoff = 20, seed = seed_num)\n",
    "\n",
    "end = time()\n",
    "print((end - sta) / 60.0, 'min')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3666b38",
   "metadata": {},
   "source": [
    "## Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "191fa846",
   "metadata": {},
   "outputs": [],
   "source": [
    "for CT in adata_query.obs['simp_name'].unique():\n",
    "    \n",
    "    print('*'*16)\n",
    "    print(CT)\n",
    "    print('*'*16)\n",
    "    print('Transfer:')\n",
    "    \n",
    "    Vis.DrawCT1(adata_query, coords_name = 'spatial_mapping',\n",
    "                CT = CT, c_name = 'simp_name',\n",
    "                root = 'Transfer2/FM_Valid2/', save = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b86dcf4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for CT in adata_query.obs['simp_name'].unique():\n",
    "    print('*'*16)\n",
    "    print(CT)\n",
    "    print('*'*16)\n",
    "    print('Transfer:')\n",
    "    Vis.DrawCT1(adata_query, coords_name = 'spatial_mapping',\n",
    "                CT = CT, c_name = 'simp_name',\n",
    "                root = 'Transfer2/FM_Valid2/', save = 'SC2HC1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "911489c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for CT in adata_ref.obs['celltype_1'].unique():\n",
    "    print('*'*16)\n",
    "    print(CT)\n",
    "    print('*'*16)\n",
    "    print('Transfer:')\n",
    "    Vis.DrawCT1(adata_ref, coords_name = 'spatial',\n",
    "                CT = CT, FM = True, c_name = 'SSV2',\n",
    "                root = 'Transfer2/FM_Valid2/', save = 'HC1')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfb59ae1",
   "metadata": {},
   "source": [
    "## Imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13eb096f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import rc\n",
    "rc('font', **{'family':'sans-serif','sans-serif':['Helvetica']})\n",
    "plt.rcParams['font.size'] = 16\n",
    "\n",
    "cmap = sns.cubehelix_palette(n_colors = 32,start = 2, rot=1.5, as_cmap = True)\n",
    "\n",
    "a = np.array([[0,1]])\n",
    "pl.figure(figsize=(9, 1.5))\n",
    "img = pl.imshow(a, cmap=cmap)\n",
    "pl.gca().set_visible(False)\n",
    "cax = pl.axes([0, 0, 0.02, 1.6])\n",
    "pl.colorbar(orientation=\"vertical\", cax=cax)\n",
    "pl.savefig(\"Transfer2/colorbar.png\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d45c6125",
   "metadata": {},
   "outputs": [],
   "source": [
    "sta = time()\n",
    "\n",
    "SI.NRD_CT(neighbors, dis, adata_ref, adata_query,\n",
    "          ct_name = 'simp_name', dis_min = 0.1, exclude_CTs = ['nan'])\n",
    "\n",
    "end = time()\n",
    "print((end - sta) / 60, 'min(s)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a6e7174",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = sns.cubehelix_palette(n_colors = 32,start = 2, rot=1.5, as_cmap = True)\n",
    "\n",
    "for CT in adata_query.obs['simp_name'].unique():\n",
    "    if(CT in ['nan']):\n",
    "        continue\n",
    "    print(CT)\n",
    "    Vis.DrawCT2(adata_ref, CT, title = True, NRD = True, colorbar = False, cmap=cmap, s=2, x='xcoord', y='ycoord',\n",
    "               root='Transfer2/FM_NRD/', save='SC2HC1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb16bfc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sta = time()\n",
    "\n",
    "adata_impute = SI.NRD_impute(neighbors, dis, adata_ref, adata_query,\n",
    "                             ct_name='simp_name', dis_min=0.1, exclude_CTs=['nan'])\n",
    "\n",
    "end = time()\n",
    "print((end - sta) / 60, 'min(s)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f635301",
   "metadata": {},
   "source": [
    "## Spatially Variable Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "721a95bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "JGs = sorted(list(set(adata_ref.var_names).intersection(set(adata_query.var_names))))\n",
    "adata = adata_ref[:, JGs]\n",
    "\n",
    "model = SI.Self_Mapping(adata, sparse = True, model_path = 'Model_SI/SI_T2.h5')\n",
    "\n",
    "SVA.PrioritizeLPG(adata, Model = model, percent = 0.5, scale_factor = 1e3,\n",
    "                  Norm = True)\n",
    "\n",
    "adata.var.sort_values('imp_sumup_norm', ascending = False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5d8beea",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = sns.cubehelix_palette(n_colors = 32,start = 2, rot=1.5, as_cmap = True)\n",
    "\n",
    "def VizCus(df, x_name, x_label, y_name, y_label, c_name = None,\n",
    "           fontsize = 28, vmin= -0.05, vmax = 0.5, s = 14, save = None):\n",
    "    \n",
    "    plt.rcParams['font.size'] = fontsize\n",
    "    plt.figure(figsize = (12, 8))\n",
    "    if(c_name == None):\n",
    "        plt.scatter(df[x_name], df[y_name], s = s)\n",
    "    elif(vmin!=None):\n",
    "        plt.scatter(df[x_name], df[y_name], s = s, alpha = 0.7,\n",
    "                c = df[c_name], vmin = vmin, vmax = vmax, cmap = cmap)\n",
    "    else:\n",
    "        plt.scatter(df[x_name], df[y_name], s = s, alpha = 0.7,\n",
    "                c = df[c_name], cmap = cmap)\n",
    "    plt.colorbar()\n",
    "    plt.xlabel(x_label)\n",
    "    plt.ylabel(y_label)\n",
    "    \n",
    "    if(save != None):\n",
    "        plt.savefig('Transfer2/' + save + '.png',\n",
    "                    dpi = 128, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "366ed2c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "JCs = sorted(list(set(adata_ref.obs_names).intersection(set(adata_impute.obs_names))))\n",
    "\n",
    "pearson_JGs = pd.DataFrame(np.zeros((len(JGs), 3)), columns = ['gene', 'pearson_r', 'p']) \n",
    "pearson_JGs['gene'] = JGs\n",
    "\n",
    "for i, JG in enumerate(JGs):\n",
    "    print(i)\n",
    "    t = stats.pearsonr(adata_impute[JCs, JG].X.flatten(), adata_ref[JCs, JG].X.toarray().flatten())\n",
    "    pearson_JGs.loc[i, 'pearson_r'] = t[0]\n",
    "    pearson_JGs.loc[i, 'p'] = t[1]\n",
    "    clear_output(wait = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a40529d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(adata, percent_top=None, log1p=False, inplace=True)\n",
    "adata.var['percent_cell'] = adata.var['n_cells_by_counts'] / adata.shape[0]\n",
    "\n",
    "t = adata.var[['percent_cell', 'imp_sumup_norm']].reset_index()\n",
    "t.columns =['gene', 'percent_cell', 'imp_sumup_norm']\n",
    "pearson_JGs = pearson_JGs.merge(t, how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e869a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "VizCus(pearson_JGs, x_name = 'imp_sumup_norm', x_label = 'Contribution to Location Prediction',\n",
    "       y_name = 'pearson_r', y_label = 'pearson_r', s = 14, c_name = 'percent_cell',\n",
    "       vmin = 0, vmax = 1, save = 'Transfer2_imp_r_pcell.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affd85d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pearson_JGs.sort_values('pearson_r', ascending = False).to_csv('Transfer2/T2_stats.csv', index = None)\n",
    "#pearson_JGs = pd.read_csv('Transfer2/T2_stats.csv')\n",
    "pearson_JGs.sort_values('pearson_r', ascending = False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0aeaee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = sns.cubehelix_palette(n_colors = 32,start = 2, rot=1.5, as_cmap = True)\n",
    "\n",
    "GeneInfo = pd.read_csv('GeneInfo_DS_CTS.csv', index_col = 0)\n",
    "\n",
    "GLs = pearson_JGs[pearson_JGs['percent_cell']>0.01].sort_values('pearson_r', ascending = False)['gene'][:20]\n",
    "    \n",
    "GLs = list(set(GLs))\n",
    "s = 4\n",
    "\n",
    "for gene in GLs:\n",
    "    print('*'*32)\n",
    "    print(gene)\n",
    "    print('*'*32)\n",
    "    print('scRNAseq:')\n",
    "    Vis.DrawGenes2(adata_query, gene = gene, lim = False, \n",
    "                   xlim = [650, 5750], ylim = [650, 5750], cmap = cmap,\n",
    "                   FM = True, CTL = None, c_name = 'simp_name', root = 'Transfer2/FM_Valid1/',\n",
    "                   s = s, x_name = 'x_transfer', y_name = 'y_transfer', title = False, save = 'AMB')\n",
    "    print('ST:')\n",
    "    Vis.DrawGenes2(adata_ref, gene = gene, lim = True,\n",
    "                   xlim = [650, 5750], ylim = [650, 5750], cmap = cmap,\n",
    "                   FM = True, CTL = None, c_name = 'SSV2', root = 'Transfer2/FM_Valid1/', \n",
    "                   s = s, x_name = 'xcoord', y_name = 'ycoord', title = False, save = 'HC1')\n",
    "    print('Imputed ST:')\n",
    "    Vis.DrawGenes2(adata_impute, gene = gene, lim = True,\n",
    "                   xlim = [650, 5750], ylim = [650, 5750], cmap = cmap,\n",
    "                   FM = True, CTL = None, c_name = 'SSV2', root = 'Transfer2/FM_Valid1/', \n",
    "                   s = s, x_name = 'xcoord', y_name = 'ycoord', title = False, save = 'HC1_Impute')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2074ab2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
